crates/egui/src/widgets/image.rs

```diff
@@ -43,15 +44,93 @@ use epaint::{util::FloatOrd, RectShape};
 #[must_use = "You should put this widget in an ui with `ui.add(widget);`"]
 #[derive(Debug, Clone)]
 pub struct Image<'a> {
-    source: ImageSource<'a>,
+    id: Id,
+    sources: ImageSources<'a>,
     texture_options: TextureOptions,
     image_options: ImageOptions,
     sense: Sense,
     size: ImageSize,
     pub(crate) show_loading_spinner: Option<bool>,
 }
 
+#[derive(Debug, Clone)]
+pub enum ImageSources<'a> {
+    Multi(Vec<(ImageSource<'a>, Duration)>),
+    Single(ImageSource<'a>),
+}
+
+impl<'a> ImageSources<'a> {
+    pub fn get_source(&self, ctx: &Context, id: Id) -> &'a ImageSource<'_> {
+        match self {
+            ImageSources::Multi(imgs) => {
+                let now = Instant::now();
+                let index = ctx.data_mut(|data| {
+                    let data = data.get_temp_mut_or_insert_with(id, || ImageData {
+                        image: 0,
+                        last_refresh: now,
+                    });
+                    let dur = imgs[data.image].1;
+                    if data.last_refresh + dur < now {
```

Comment:
We shouldn't need to store any state. Just calculate the sum of the frame time, then modulate the current time with that and calculate the frame index and how long until the next frame. No need for `id`